src/software/ai/hl/stp/tactic/goalie_tactic.cpp

```diff
@@ -71,67 +150,145 @@ void GoalieTactic::calculateNextIntent(IntentCoroutine::push_type &yield)
+            }
+            // if the ball is slow or stationary inside our defense area, and is safe
+            // to do so, chip it out
+            else
+            {
+                next_intent = chip_action.updateStateAndGetNextIntent(
+                    *robot, ball, ball.position(),
+                    (ball.position() - field.friendlyGoal()).orientation(), 2);
+            }
         }
-        // Case 3
+        // case 3: ball does not have a clear velocity vector towards the goal, so
+        // position goalie in best position to block shot
         else
         {
-            // block the cone by default
-            float radius =
-                Util::DynamicParameters::GoalieTactic::block_cone_buffer.value() +
-                ROBOT_MAX_RADIUS_METERS;
-            Point goalie_pos =
-                calcBlockCone(field.friendlyGoalpostNeg(), field.friendlyGoalpostPos(),
-                              ball.position(), radius);
-
-            // restrict the goalie to a semicircle inscribed inside the defense area
-            Point goalie_restricted_pos =
-                field.friendlyGoal() - (field.friendlyDefenseArea().yLength() *
-                                        (field.friendlyGoal() - goalie_pos).norm());
-
-            // restrict the point to be within the defense area
-            auto goalie_orientation = (ball.position() - goalie_pos).orientation();
-            next_intent             = move_action.updateStateAndGetNextIntent(
-                *robot, goalie_restricted_pos, goalie_orientation, 0.0, false, false,
+            // compute angle between two vectors, negative goal post to ball and positive
+            // goal post to ball
+            Angle block_cone_angle =
+                (ball.position() - field.friendlyGoalpostNeg())
+                    .orientation()
+                    .minDiff(
+                        (ball.position() - field.friendlyGoalpostPos()).orientation());
+
+            // compute block cone position, allowing 1 ROBOT_MAX_RADIUS_METERS extra on
+            // either side
+            Point goalie_pos = calcBlockCone(
+                field.friendlyGoalpostNeg(), field.friendlyGoalpostPos(), ball.position(),
+                block_cone_radius * block_cone_angle.toRadians());
+
+            // we want to restrict the block cone to the friendly crease, also potentially
+            // scaled by a defense_area_deflation_parameter
+            Rectangle deflated_defense_area = field.friendlyDefenseArea();
+            deflated_defense_area.expand(-defense_area_deflation);
+
+            // restrain the goalie in the deflated defense area, if the goalie cannot be
+            // restrained or if there is no proper intersection, then we safely default to
+            // center of the goal
+            auto clamped_goalie_pos =
+                restrainGoalieInRectangle(goalie_pos, deflated_defense_area);
+
+            if (!clamped_goalie_pos)
```

Comment:
I think there should be a comment here explaining what you're doing with the goalie when there is no clamped position

Reply:
added a comment